mender-update/context.hpp

```diff
@@ -68,6 +68,9 @@ class MenderContext {
 	ExpectedProvidesData LoadProvides();
 	expected::ExpectedString GetDeviceType();
 	error::Error CommitArtifactData(const ProvidesData &data);
+	const conf::MenderConfig &GetConfig() const {
```

Comment:
Q: I seem to recall someone having an argument with me about returning const references, or is this different?

Reply:
Yes, it's different. These are generally the recommendations:

* **Return const by reference**: Usually the best choice when you are returning a member variable, because you get the best performance, and the const prevents the caller from modifying your internals. If the caller assigns to a variable which is non-const, then a copy is made, and the caller won't even know that the return was originally a reference.
* **Return non-const by reference**: Also ok, if, and only if, the caller is allowed to modify the object directly. Useful for informational structs, for example, but still a far less common return type than const.
* **Return const by value**: Should never be used. Since it's by value, the object is owned by the caller, and hence it is a bad idea to enforce const, because you don't care what the caller does to it. And in fact, you cannot enforce const, because if the caller wants non-const, it will just copy-construct the object into a new non-const object. So you get an extra copy instead of move semantics. This is the one I pointed out in your PR earlier.
* **Return non-const by value**: The only appropriate choice when you are returning something you constructed in the function being called, in which case references cannot be used.